File: Khelmianov-Artem/Term4/Block3/Dekanat/src/main/kotlin/collections/PhasedCuckooHashSet.kt
package collections

import java.util.concurrent.atomic.AtomicInteger

abstract class PhasedCuckooHashSet<T : Any>(
    @Volatile protected var capacity: Int
) : ConcurrentSet<T> {
    @Volatile
    protected var storage = Array(2) { Array(this.capacity) { mutableListOf<T>() } }

    protected val THRESHOLD = 16
    protected val LIST_SIZE = 32
    protected val RELOCATE_LIMIT = 5

    override val size: Int get() = _size.get()
    protected val _size = AtomicInteger(0)

    protected abstract fun acquire(element: T);
    protected abstract fun release(element: T);
    protected abstract fun resize();

    protected inline fun <R> withLock(element: T, block: () -> R): R {
        acquire(element)
        try {
            return block()
        } finally {
            release(element)
        }
    }

    override fun add(element: T): Boolean {
        var i: Int = 0
        var h: Int = 0
        var mustResize = false;

        withLock(element) {
            val h0 = element.hash(0)
            val h1 = element.hash(1)

            if (contains(element)) return false
            val s0 = storage[0, h0]
            val s1 = storage[0, h1]

            if (s0.size < THRESHOLD) {
                _size.incrementAndGet()
                s0.add(element)
                return true
            } else if (s1.size < THRESHOLD) {
                _size.incrementAndGet()
                s1.add(element)
                return true
            } else if (s0.size < LIST_SIZE) {
                i = 0
                h = h0
                _size.incrementAndGet()
                s0.add(element)
            } else if (s1.size < LIST_SIZE) {
                i = 1
                h = h1
                _size.incrementAndGet()
                s1.add(element)
            } else {
                mustResize = true
            }
        }
        if (mustResize) {
            resize()
            add(element)
        } else if (!relocate(i, h)) {
            resize()
        }
        return true
    }

    override fun remove(element: T): Boolean = withLock(element) {
        return if (storage[0, element.hash(0)].remove(element)) {
            _size.decrementAndGet()
            true
        } else if (storage[1, element.hash(1)].remove(element)) {
            _size.decrementAndGet()
            true
        } else false
    }

    override fun contains(element: T): Boolean = withLock(element) {
        if (element in storage[0, element.hash(0)]) true
        else element in storage[1, element.hash(1)]
    }

    @Suppress("NAME_SHADOWING")
    protected fun relocate(i: Int, hi: Int): Boolean {
        var i = i
        var hi = hi
        var hj: Int;
        var j = 1 - i;
        repeat(RELOCATE_LIMIT) {
            val iSet = storage[i, hi]
            val y = iSet/*[0]*/.getOrNull(0) ?: return@repeat
            hj = y.hash(j)
            withLock(y) {
                val jSet = storage[j, hj]
                if (iSet.remove(y)) {
                    if (jSet.size < THRESHOLD) {
                        jSet.add(y)
                        return true
                    } else if (jSet.size < LIST_SIZE) {
                        jSet.add(y)
                        i = 1 - i
                        hi = hj
                        j = 1 - j
                    } else {
                        iSet.add(y)
                        return false
                    }
                } else if (iSet.size < THRESHOLD) {
                    return true
                }
            }
        }
        return false
    }

    protected fun T.hash(array: Int, modulus: Int = capacity): Int {
        return if (array == 0)
            this.hashCode() % modulus
        else
            (this.hashCode() / modulus) % modulus
    }
}

internal operator fun <A> Array<Array<A>>.get(x: Int, y: Int): A = this[x][y]
internal fun <A> Array<Array<A>>.forEach(action: (A) -> Unit): Unit = forEach { it: Array<A> ->
    it.forEach { action(it) }
}

File: Arthur Alekseev/Term 4/Task 3.1/ExamSystem/ExamSystem.Core/Sets/StripedCuckooHashSet/PhasedCuckooHashSet.cs
namespace ExamSystem.Core.Sets.PhasedCuckooHashSet;

public abstract class PhasedCuckooHashSet<T> : Interface.ISet<T>
{
	// list is semi-full
	private const int Threshold = 100;

	// list is full
	protected const int ListSize = 128;

	// steps to relocate
	private const int Limit = 40;

	private const int HashStride = 0x2345;

	protected volatile int Capacity;
	protected volatile List<T>[,] Table;

	protected PhasedCuckooHashSet(int size)
	{
		Capacity = size;
		Table = new List<T>[2, Capacity];

		for (var i = 0; i < 2; i++)
		for (var j = 0; j < Capacity; j++)
			Table[i, j] = new List<T>(ListSize);
	}

	public bool Contains(T x)
	{
		Acquire(x);
		try
		{
			var set0 = Table[0, Hash0(x) % Capacity];

			if (set0.Contains(x))
				return true;

			var set1 = Table[1, Hash1(x) % Capacity];
			return set1.Contains(x);
		}
		finally
		{
			Release(x);
		}
	}

	public bool Remove(T x)
	{
		Acquire(x);
		try
		{
			var set0 = Table[0, Hash0(x) % Capacity];
			if (set0.Contains(x))
			{
				set0.Remove(x);
				return true;
			}

			var set1 = Table[1, Hash1(x) % Capacity];

			if (!set1.Contains(x))
				return false;

			set1.Remove(x);

			return true;
		}
		finally
		{
			Release(x);
		}
	}

	public bool Add(T x)
	{
		Acquire(x);
		var h0 = Hash0(x) % Capacity;
		var h1 = Hash1(x) % Capacity;

		var i = -1;
		var h = -1;

		var mustResize = false;

		try
		{
			if (Contains(x))
				return false;

			var set0 = Table[0, h0];
			var set1 = Table[1, h1];

			if (set0.Count < Threshold)
			{
				set0.Add(x);
				return true;
			}
			else if (set1.Count < Threshold)
			{
				set1.Add(x);
				return true;
			}
			else if (set0.Count < ListSize)
			{
				set0.Add(x);
				i = 0;
				h = h0;
			}
			else if (set1.Count < ListSize)
			{
				set1.Add(x);
				i = 1;
				h = h1;
			}
			else
			{
				mustResize = true;
			}
		}
		finally
		{
			Release(x);
		}

		if (mustResize)
		{
			Resize();
			Add(x);
		}
		else if (!Relocate(i, h))
		{
			Resize();
		}

		return true;
	}

	protected int Hash0(T item)
	{
		if (item is null)
			throw new ArgumentNullException(nameof(item));

		return Math.Abs(item.GetHashCode());
	}

	protected int Hash1(T item)
	{
		if (item is null)
			throw new ArgumentNullException(nameof(item));

		return Math.Abs(item.GetHashCode() + HashStride);
	}

	protected abstract void Acquire(T x);
	protected abstract void Release(T x);
	protected abstract void Resize();

	private bool Relocate(int i, int hi)
	{
		var hj = 0;
		var j = 1 - i;
		for (var round = 0; round < Limit; round++)
		{
			var iSet = Table[i, hi];
			var y = iSet[0];

			hj = i switch
			{
				0 => Hash1(y) % Capacity,
				1 => Hash0(y) % Capacity,
				_ => hj
			};

			Acquire(y);

			var jSet = Table[j, hj];

			try
			{
				if (iSet.Remove(y))
					switch (jSet.Count)
					{
						case < Threshold:
							jSet.Add(y);
							return true;

						case < ListSize:
							jSet.Add(y);
							i = 1 - i;
							hi = hj;
							j = 1 - j;
							break;

						default:
							iSet.Add(y);
							return false;
					}
				else if (iSet.Count < Threshold) return true;
			}
			finally
			{
				Release(y);
			}
		}

		return false;
	}
}